src/cuda/cuda_sampling.cu

```diff
@@ -431,53 +439,178 @@ __global__ void FilterOnTopP(float* scores, float* prefix_sums, float* scores_te
+      partial.insert(elem, elemId);
+    }
+    // reduce in thread block
+    TopK_2 top_k_sequence = BlockReduce(temp_storage).Reduce(partial, reduce_topk_op_2);
+
+    if (tid == 0) {
+      distributed_scores_out_curr[ite + batch * k] = top_k_sequence.u;
+      distributed_indices_out_curr[ite + batch * k] = top_k_sequence.p;
+
+      // set the ax value to -MAX_T_VAL so that the value doesn't get picked again
+      scores_in[batch * vocab_size + top_k_sequence.p] = -MAX_T_VAL;
+    }
+
+    __syncthreads();
+  }
+
+  // All TBs flush their data and it should be visible to every other TB
+  __threadfence();   
   __syncthreads();
 
-  if (threadIdx.x == 0) {
-    for (int i = 0; i < k; i++) {
-      scores_out[i + batch * k] = top_k_sequence.value[i] / temperature;
-      indices_out[i + batch * k] = top_k_sequence.key[i];
+  // Signal that each threadblock has done its work
+  if (threadIdx.x == 0)
+  {
+    atomicAdd(top_k_distributed_lock, 1);
+  }
+
+  // The reduction threadbloc's elected thread spins while waiting
+  if (blockIdx.z == 0) {
+    if (threadIdx.x == 0)
+    {
+      uint count_of_completed_TBs = 0;
+
+      asm volatile("ld.volatile.global.u32 %0, [%1];" : "=r"(count_of_completed_TBs) : "l"(top_k_distributed_lock));
+      while (count_of_completed_TBs < top_k_shards)
+      {
+          asm volatile("ld.volatile.global.u32 %0, [%1];" : "=r"(count_of_completed_TBs) : "l"(top_k_distributed_lock));
+      }
+
+      // Reset the lock
+      atomicExch(top_k_distributed_lock, 0);
     }
+
+    TopK_2 reduce_partial; 
+    for (int ite = 0; ite < k; ite++) {
+      reduce_partial.init();
+  
+      for (int i = threadIdx.x; i < top_k_shards*k; i += blockDim.x) {
+        float score = distributed_scores_out[i];
+        int index = distributed_indices_out[i];
+        reduce_partial.insert(score, index);
+      }
+
+      // reduce in thread block
+      TopK_2 top_k_sequence_reduced = BlockReduce(temp_storage).Reduce(reduce_partial, reduce_topk_op_2);
+
+      if (tid == 0) {
+        scores_out[ite + batch * k] = top_k_sequence_reduced.u / temperature;
+        indices_out[ite + batch * k] = top_k_sequence_reduced.p;
+      }
+
+      __syncthreads();
+    }
+
   }
 }
 
 // Gets all top K indices and scores from unsorted input
-template <int max_k>
-void LaunchGetTopKSubset(cudaStream_t stream, float* scores_in, float* scores_out, int* indices_out, int vocab_size, int batch_size, int k, float temperature) {
-  dim3 grid(batch_size, 1, 1);
-  dim3 block(256, 1, 1);
-  GetTopKKernel<max_k, 256><<<grid, block, 0, stream>>>(indices_out, scores_in, scores_out, batch_size, vocab_size, k, temperature);
+void LaunchGetTopKSubset(cudaStream_t stream, float* scores_in, float* scores_out, int* indices_out, int vocab_size, int batch_size, int k, float temperature, SamplingData* data = nullptr) {
+  bool enable_distributed_top_k = true;
+  if (batch_size == 1 && k <= 64 && enable_distributed_top_k) {
```

Comment:
Right now it can only supported k <= 50 (this is a bug) - I will send out a fresh PR based on this once folks are happy with the implementation

Reply:
Fixed